src/pudl/transform/vcerare.py

```diff
@@ -0,0 +1,330 @@
+    return df
+
+
+def _check_for_valid_counties(
+    df: pd.DataFrame, clean_fips_df: pd.DataFrame, df_name: str
+) -> pd.DataFrame:
+    """Make sure the state_county values show up in the fips table.
+
+    This step happens before the table gets stacked to save memory.
+    """
+    logger.info(f"Checking for valid counties in the {df_name} table.")
+    county_state_names_fips = clean_fips_df.county_state_names.unique().tolist()
+    county_state_names_cap_fac = df.columns.tolist()
+    expected_non_counties = ["report_year", "unnamed_0"]
+    non_county_cols = [
+        x
+        for x in county_state_names_fips
+        if x not in county_state_names_cap_fac + expected_non_counties
+    ]
+    if non_county_cols:
+        raise AssertionError(f"""found unexpected columns that aren't in the FIPS table:
+{non_county_cols}.""")
+    return df
+
+
+def _combine_all_cap_fac_dfs(cap_fac_dict: dict[str, pd.DataFrame]) -> pd.DataFrame:
+    """Combine capacity factor tables."""
+    logger.info("Merging all the capacity factor tables into one")
+    merge_keys = ["report_year", "datetime_utc", "hour_of_year", "county_state_names"]
+    for name, df in cap_fac_dict.items():
+        cap_fac_dict[name] = df.set_index(merge_keys)
+    mega_df = pd.concat(
+        [
+            cap_fac_dict["solar_pv"],
+            cap_fac_dict["offshore_wind"],
+            cap_fac_dict["onshore_wind"],
+        ],
+        axis="columns",
+    ).reset_index()
+
+    return mega_df
+
+
+def _combine_cap_fac_with_fips_df(
+    cap_fac_df: pd.DataFrame, fips_df: pd.DataFrame
+) -> pd.DataFrame:
+    """Combine the combined capacity factor df with the fips df."""
+    logger.info(
+        "Merging the combined capacity factor table with the Lat-Long-FIPS table"
+    )
+    combined_df = pd.merge(cap_fac_df, fips_df, on="county_state_names", how="left")
+    return combined_df
+
+
+def _combine_city_county_records(df: pd.DataFrame) -> pd.DataFrame:
+    """Average cap fac values for city and county values with the same FIPS ID.
+
+    There are several duplicate FIPS code values in the data. Most of them
+    are lakes within a county, but some of them are cities. Only one of the
+    cities, clifton forge city, has values, so we average those values
+    with that of the county. The other city, XXX, has no values so we
+    remove it.
+
+    """
+    bad_county_mask = df["county_state_names"].isin(
+        ["alleghany_virginia", "clifton_forge_city_virginia"]
+    )
+    # Take the subset of the data that's relevant to the changes and sort the values
+    # by county_state so that alleghany is first and those are the values that are
+    # preserved by the groupby "first" below.
+    subset_df = df[bad_county_mask].sort_values(by=["county_state_names"])
+    cap_fac_cols = [x for x in subset_df.columns if "capacity_factor" in x]
+    other_cols = [
+        x for x in subset_df.columns if x not in cap_fac_cols + ["datetime_utc"]
```

Comment:
Is `datetime_utc` not in `subset_df.columns`? Just curious why this is necessary.

Reply:
`datetime_utc` is in `subset_df.columns` but datetime_utc becomes the index value when we run the groupby so we don't want to included in the list of `other_cols` which is used to take the "first" value (vs. the cap fac values which use "mean") when the data are aggregated. I can try and make this more clear with some inline comments.